### DISCUSSION TITLE
Cron Job runs, hits correct route, and is showing it received a 200, but nothing in the route it hits runs.

### DISCUSSION THREAD

**Discussion by PlasticPrintersMN** [vercel/community#3903] [Help] [answered]
### Summary

I am on a pro plan. The cron I set up is set to work every 2 minutes. The very first time it runs it hits the correct route, everything in the route works including updating my db, and shows the correct console logs.

Every time it runs after it says it hits the correct route, and shows a 200 response, but nothing runs in the route itself. No logs, no db updates. Hoping it's just something I'm overlooking. Adding some screen shots for reference:

Here is vercel.json:
![image](https://github.com/vercel/community/assets/51180935/eb3bc582-1209-43e9-9273-020f4f259717)

Here is the route in the Explorer:
![image](https://github.com/vercel/community/assets/51180935/8121c240-b900-4389-8f7d-86164093b07a)

Here is the route itself:
![image](https://github.com/vercel/community/assets/51180935/e2ade066-0169-4436-a017-b9c2bf72038f)

Here is the first time the cron ran including 200 response and console logs:
![image](https://github.com/vercel/community/assets/51180935/a60d5231-1326-4435-b85c-d7f1842b3a80)

And here is the Cron continuing to run and returning 200, but not getting any console logs:
![image](https://github.com/vercel/community/assets/51180935/10bb04ad-3d3d-4905-9dd4-fdaac6bdc1f3)

A couple things I noticed is for some reason it seems to grab the route twice each time. Once is logged as a "Serverless Function" and once as a "Request". Not sure what that is about. And second, all of the logs show the environment is preview despite not having any preview builds at all. Only production builds:

![image](https://github.com/vercel/community/assets/51180935/73b797c8-0500-4798-b423-a4bb94c08c54)

![image](https://github.com/vercel/community/assets/51180935/f6a9905e-ae14-418f-acce-d28d3cb547b6)

Here is the version of Next I am running:

"next": "13.4.19"

Hopefully this is enough information for someone to point me in the right direction. 

Thank you so much for your time.

### Steps to Reproduce

Set up a cron. Deploy to production. Let it run. It only runs the function the first time.

**Comment by PlasticPrintersMN** [comment]
It's so weird, it is showing that all of them are running in the Preview environment:

![image](https://github.com/vercel/community/assets/51180935/36c35122-d406-4327-8041-5da166a4bc81)

I don't have any Preview builds for one, and two I thought crons were Production only...

**Reply by PlasticPrintersMN** [reply]
This has to be the issue. I just hit the deployed route from post man and everything ran perfectly. Including the console logs and the db updates. That log even showed Production environment:

![image](https://github.com/vercel/community/assets/51180935/cc18a19b-bf4f-46f9-af66-7da9509def09)

here is the log:

![image](https://github.com/vercel/community/assets/51180935/807b623c-bc4e-4789-a58b-f65a2054fc7c)

**Reply by PlasticPrintersMN** [reply]
But why is the cron running in the Preview environment?

**Comment by Rrhapsod** [comment]
I have the same problem, the cron jobs running on preview and not running the code inside. I'm on Hobby plan.

**Comment by NickStull** [comment]
Facing the same issue. Can't seem to get the cron to run in the production environment despite that being the only place it's supposed to run.

**Comment by evanfurniss** [comment]
I encountered the same thing. Any word on a resolution?

**Comment by PlasticPrintersMN** [comment]
Have there been any eyes on this yet?

**Comment by chris-nowicki** [comment, marked as answer]
@PlasticPrintersMN I had this issue a while back when my cron jobs stoped working.  Look at your vercel build log and is your Cron Job route listed as static or server component.  Static would be indicated by a circle next to the route.  If it is a static route then try adding a route segment config to force the route to be dynamic (server) route.  The code should go before the route function:

```js
// Nextjs route segment config
export const dynamic = 'force-dynamic' // Force dynamic (server) route instead of static page
```

What **MAY** be happening is at build time the route is generated as a static page with the data at that time.  Then when the cron job runs it hits the data that was generated at build time, indicated by the 200 success, but data does **NOT** change because it keeps hitting the same static data rather than being dynamic and getting data _EACH_ time the route is hit by the cron job... again this depends on what your build log says.

**Reply by PlasticPrintersMN** [reply]
This 100% fixed the issue. Man, this is not very clear at all. I still think there is some type of bug being that the logs were pointing toward some non-existent preview environment, and all of their docs say cron is prod only.

Either way. Thank you so much for the help. A week later and Vercel can't be bothered to respond to a Pro Accounts support ticket. Gross.

**Reply by PlasticPrintersMN** [reply]
@Rrhapsod

**Reply by amyegan** [reply]
Thanks for sharing this answer! For anyone who wants more info, you can find it in the [Next.js Route Handlers](https://nextjs.org/docs/app/building-your-application/routing/route-handlers) documentation at https://nextjs.org/docs/app/building-your-application/routing/route-handlers#segment-config-options

**Reply by amyegan** [reply]
Thank you all for reporting this. I asked the engineering team for more information and help with this problem. 

Regarding [route segment config](https://nextjs.org/docs/app/building-your-application/routing/route-handlers#segment-config-options), route handlers follow the same strategy as pages. By default, a static build of the route is attempted and is switched as soon as a dynamic dependency is detected. You can change this using one of the [route segment config options](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config).

If you rely on request headers, request URL, or do a non-static fetch all of these can change per-request and should be treated as dynamic. If your function is set up in a way that should be interpreted as dynamic but it still turns into a static route, please share the function code so I can pass it on to the Next.js engineers. 🙏 

Regarding the unexpected Cron Jobs `preview` vs `production` behavior, I don't have an update to share yet. But the engineering team is investigating and I will share new information once I have it.

**Reply by louisthomaspro** [reply]
Any updates @amyegan ?
My production cron jobs are calling the deployment preview link (`<project-name>-<unique-hash>-<scope-slug>.vercel.app`) instead of my registered domain...

**Reply by amyegan** [reply]
The team is currently testing a fix for the Cron Jobs `preview` vs `production` mislabeling.

@louisthomaspro All deployments, including `production`, have a unique [generated URL](https://vercel.com/docs/deployments/generated-urls#). It is expected behavior that the cron job uses the unique URL rather than the custom domain for you project. This avoids conflicts as new deployments could happen at any time and would disrupt running jobs if they followed the custom domain.

**Reply by louisthomaspro** [reply]
Thanks for your answer. It makes sense!
I ran into a different issue then. I want the cron job to execute `revalidateTag('test')`, but it's not working. I thought is was because of the domain. Any idea why ?
I had to do this workaround: the cron is calling an API endpoint with the right domain (ex: https://registered-domain/api/revalidate)

**Comment by Rrhapsod** [comment]
> @Rrhapsod

Yeah, I called Chris from the page dev.to and he is a savior =D

Amazing how a single line that could be in the manuals can solve the problem in an easy way
